design/bios-config.md

```diff
@@ -0,0 +1,151 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# bios-config
+
+## Status
+
+provisional
+
+## Table of Contents
+
+<!--ts-->
+   * [Title](#title)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives [optional]](#alternatives-optional)
+      * [References](#references)
+
+<!-- Added by: stack, at: 2019-02-15T11:41-05:00 -->
+
+<!--te-->
+
+## Summary
+
+This document explains the usage of YAML attributes for BIOS configuration for the BareMetalHost of the BMO. These configurations will be applied to the host according to the vendor's BMC driver being used by Ironic.
+
+## Motivation
+
+We need this document to specify the typical usage of BIOS YAML attributes in the BareMetalHost YAML. 
+
+
+### Goals
+
+1. To agree on the format of the attributes that are to be supported as a starting point
+2. To keep the attribute naming vendor agnostic
+
+### Non-Goals
+
+1. To list every attribute for every vendor exhaustively
+
+## Proposal
+The proposal is to set/unset the BIOS configuration values using vendor drivers available for each BMC type. For this purpose, new YAML attributes are to be introduced in the BareMetalHost spec. 
+ 
+### Implementation Details/Notes/Constraints 
+All the BIOS related attributes or fields will come under the sub-section called ```bios``` in spec. The values for each attribute can be a boolean or a string. The vendor driver type to implement this configuration is not to be specified separetely as it is already known from the bmc sub-section.
+Initially, the BIOS configuration can be applied using the controller types idrac and irmc
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+ name: bm0
+spec:
+ online: true
+ bmc:
+   address: idrac://192.168.122.1:6230/
+   credentialsName: bm0-bmc-secret
+ bootMACAddress: 52:54:00:b7:b2:6f	
+ bios:
+   EnableSriovGlobal: false
+   BootMode: uefi
```

Comment:
Let's not mix `BootMode` in this picture, it has wider implications on ironic and must be handled separately.

Reply:
Can you kindly elaborate on these implications? And under these implications, what would be a recommended way to handle ```BootMode```?

Reply:
Boot mode has driver behavior implications as in it is a state that has to be asserted for some hardware in different workflow steps. The best example I can think of off the top of my head is when IPMI is used. When you want to set the machine to boot UEFI, you need to send along a raw command to the BMC, as saying "boot network" or "boot disk" implies bios mode. IPMI of course being last revised before UEFI was generally available, if I'm remembering dates correctly.

Reply:
In case of Redfish with iDRAC BMC, the BootMode can be set simply as an attribute ```BootMode```. I wasn't aware of the way IPMI handles this, the proposal is based on iDRAC configurations. 
Since we are to pursue a generic representation for configurations, how do you propose we depict them here? In this file? As @dtantsur says, it should be handled separately, can it be implied to mean that the ```BootMode``` tag should be placed elsewhere in the YAML? If so, where?

Reply:
Alternatively, I think we can discuss BootMode parameter later. I can make the necessary edits to include the BIOS configurations that can be applied uniformly (in the same way) across all hardware types for Ironic.

Reply:
Ironic has to be aware of the boot mode a node uses. If you set `BootMode=UEFI` here, but forget to tell ironic via `properties`, ironic will prepare an incorrect boot environment (assuming legacy boot).

I think we should have it as a top-level property. If the iDRAC driver wants to be able to change it, it has to implement this interface: https://opendev.org/openstack/ironic/src/branch/master/ironic/drivers/base.py#L896-L959

Reply:
I expected "boot mode" to be supported by all of the ironic drivers. Is that not the case?

Reply:
It's a bit more complicated. Every driver supports adapting to boot mode, as in "this node uses boot mode UEFI, please adjust accordingly". What not every driver supports is "I want this node to be in UEFI, deal with it". IPMI only supports the former. Supporting the latter requires writing some explicit code.

Reply:
I think ```BootMode``` can be handled separately and is out of scope of BIOS configurations. It's also been defered to be implemented later. I will make the edits to the proposal.